mitodl/mit-open
# Add LearningPath (StaffList) model and API, refactor relationships between resources-

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #13
Closes #16
Closes #32
Closes #51
Partially addresses #19 

#### What's this PR do?
- Refactors relationships between learning resources
- Adds LearningPath (formerly StaffList) model and a CRUD API for it.
- Adds Django Admin functionality for learning_resources

#### How should this be manually tested?
If you haven't already, run the following:

```
docker compose run --rm web python manage.py loaddata platforms departments offerers
docker compose run --rm web python manage.py backpopulate_xpro_data
```

Go to the following urls, they should return data that looks reasonable.  In particular, the programs returned should have a "courses" field containing a list of course LearningResource objects
`http://localhost/api/v1/learning_resources`
`http://localhost/api/v1/courses`
`http://localhost/api/v1/courses/<learning_resource_id>/`
`http://localhost/api/v1/programs`
`http://localhost/api/v1/programs/<learning_resource_id>/`

LearningPath CRUD
- Go to `http://localhost/api/v1/learningpaths/` as an admin user
- Submit the following JSON in the raw data field:
```json
{
    "title": "foo",
    "description": "bar",
    "topics": [1, 2],
    "published": true
}
```
- A new LearningPath resource should be created
- Go to the resources endpoint for it `http://localhost:8063/api/v1/learningpaths/<path_resource_id>/resources/`
- Submit the following JSON in the raw data field, using the id of a course LearningResource:
```json
{
    "child": <course_learning_resource_id>
}
```
That course should then show up in the above url response with a position of 1.  Repeat with another course, it should appear in the response with a position of 2

- Go to `http://localhost:8063/api/v1/learningpaths/<path_resource_id>/resources/<item_id>/` and submit the following JSON:
```json
{
    "position": 99
}
```
The position value should be updated.
- Go to `http://localhost:8063/api/v1/learningpaths/<path_resource_id>/resources/<item_id>/` and click the delete button, that item should be removed.
- Go to `http://localhost:8063/api/v1/learningpaths/<path_resource_id>/` and click the delete button, that learning path should be removed.


#### Any background context you want to provide?
Some differences between this implementation of `LearningPath` and the old `StaffList`:

- I did not include a`privacy_level` field ("public" or "private") on `LearningPath` because `LearningResource.published` (true/false) can be used to make that distinction.
- Also removed the distinction between a "list" (unsortable) and "path" (sortable) - a LearningPath will always be sortable.
- I did not implement `UserList` in this PR because after some discussion w/Ferdi it was decided that `UserList` should not have a foreign key to `LearningResource`.  It will have its own fields for title, description, etc and have its own many-to-many relationship to other learning resources.
